plugin/trino-password-authenticators/src/main/java/io/trino/plugin/password/ldap/LdapGroupProvider.java

```diff
@@ -0,0 +1,219 @@
+        requireNonNull(ldapConfig, "ldapConfig is null");
+
+        this.userBindSearchPatterns = ldapConfig.getUserBindSearchPatterns();
+        this.userBaseDistinguishedName = Optional.ofNullable(ldapConfig.getUserBaseDistinguishedName());
+        this.groupBaseDistinguishedName = ldapConfig.getGroupBaseDistinguishedName();
+        this.bindDistinguishedName = ldapConfig.getBindDistinguishedName();
+        this.bindPassword = ldapConfig.getBindPassword();
+        this.allowUserNotExist = ldapConfig.isAllowUserNotExist();
+        this.groupSearchFilter = ldapConfig.getGroupSearchFilter();
+        this.groupCache = buildNonEvictableCacheWithWeakInvalidateAll(
+                CacheBuilder.newBuilder()
+                        .expireAfterWrite(ldapConfig.getLdapCacheTtl().toMillis(), MILLISECONDS),
+                CacheLoader.from(this::getGroupsWithBindDistinguishedName));
+    }
+
+    @VisibleForTesting
+    void invalidateCache()
+    {
+        groupCache.invalidateAll();
+    }
+
+    @Override
+    public Set<String> getGroups(String user)
+    {
+        try (ThreadContextClassLoader ignored = new ThreadContextClassLoader(getClass().getClassLoader())) {
+            return groupCache.getUnchecked(user);
+        }
+        catch (UncheckedExecutionException e) {
+            if (e.getCause() instanceof AccessDeniedException) {
+                if (allowUserNotExist) {
+                    return Collections.emptySet();
+                }
+                else {
+                    throw (AccessDeniedException) e.getCause();
+                }
+            }
+            throw e;
+        }
+    }
+
+    private Set<String> getGroupsWithBindDistinguishedName(String user)
+    {
+        Set<String> groupSet = Collections.emptySet();
+
+        if (containsSpecialCharacters(user)) {
+            throw new AccessDeniedException("Username contains a special LDAP character");
+        }
+
+        try {
+            for (String userBindSearchPattern : userBindSearchPatterns) {
+                String searchFilter = replaceUser(userBindSearchPattern, user);
+                String userDistinguishedName = lookupUserDistinguishedName(
+                        this.userBaseDistinguishedName.orElseThrow(),
+                        searchFilter,
+                        this.bindDistinguishedName,
+                        this.bindPassword);
+                String groupFilter = replaceUser(this.groupSearchFilter, userDistinguishedName);
+                log.debug("Querying LDAP from %s using filter %s", this.groupBaseDistinguishedName, groupFilter);
+                Set<String> queryResult = lookupUserDistinguishedNames(
+                        this.groupBaseDistinguishedName, groupFilter,
+                        bindDistinguishedName, bindPassword);
+                groupSet = queryResult.stream().map(source -> {
+                    try {
+                        log.debug("extracting group from %s", source);
```

Comment:
Instead of matching pattern can we fetch the actual attributes ?

Reply:
What attributes are you talking about? Depending on what they are, we can add more debug logs if needed.